lib/statsd.rb

```diff
@@ -196,16 +211,16 @@ def send_to_socket(message)
     def read_from_socket
       buffer = ""
       loop do
-        line = socket.readline
+        line = @socket.readline
         break if line == "END\n"
         buffer += line
       end
-      socket.readline # clear the closing newline out of the socket
+      @socket.readline # clear the closing newline out of the socket
       buffer
     end
 
-    def socket
-      Thread.current[:statsd_admin_socket] ||= TCPSocket.new(host, port)
+    def connect
+      @s_mu.synchronize { @socket = TCPSocket.new(host, port) }
```

Comment:
I would suggest removing the `connect` call form `host=` and `port=`. If the host at the port change then there is no way to change them together.